src/v/security/gssapi_principal_mapper.cc

```diff
@@ -0,0 +1,268 @@
+  int number_of_components,
+  std::string_view format,
+  std::string_view match,
+  std::string_view from_pattern,
+  std::string_view to_pattern,
+  repeat repeat_,
+  make_lower to_lower,
+  make_upper to_upper)
+  : _default_realm(default_realm)
+  , _is_default(false)
+  , _number_of_components(number_of_components)
+  , _format(format)
+  , _match(match)
+  , _from_pattern(std::regex{
+      from_pattern.begin(),
+      from_pattern.length(),
+      std::regex_constants::ECMAScript | std::regex_constants::optimize})
+  , _from_pattern_str(from_pattern)
+  , _to_pattern(to_pattern)
+  , _repeat(repeat_)
+  , _to_lower(to_lower)
+  , _to_upper(to_upper) {}
+
+std::optional<ss::sstring>
+gssapi_rule::apply(std::vector<std::string_view> params) const {
+    static const RE2 non_simple_regex(non_simple_pattern, RE2::Quiet);
+    const re2::StringPiece match_piece(_match.data(), _match.size());
+    const RE2 match_regex(match_piece, RE2::Quiet);
+    vassert(
+      non_simple_regex.ok(),
+      "Invalid non-simple-regex: {}",
+      non_simple_regex.error());
+    // Even if _match is empty, the RE2 will be 'ok'
+    vassert(match_regex.ok(), "Invalid match regex: {}", match_regex.error());
+
+    /*
+     * How this works:
+     *
+     * Take a rule like this:
+     *
+     * RULE:[2:$1foo$2](Admin\.(.*)s/Admin\.(.*)/$1/
+     *
+     * And say the GSSAPI principal name is:
+     * Admin.site-user/example.com@REALM.com.
+     *
+     * The name breaks down to:
+     * service-name: Admin.site-user
+     * host-name: example.com
+     * realm: REALM.com
+     *
+     * First, the "2" at the beginning ensures there are two parameters in the
+     * principal name (ignoring realm, we have the service-name and host-name).
+     *
+     * If we were missing the host-name, then we would only have one parameter
+     * and this rule would be ignored.
+     *
+     * Now, the "$1foo$2".  This is the 'format'.  This formats the GSSAPI name,
+     * so it can be fed into the next match regex.
+     *
+     * $0 - Realm
+     * $1 - service-name
+     * $2 - host-name
+     *
+     * So this 'format' takes the name and turns it into
+     * "Admin.site-userfooexample.com"
+     *
+     * Next, we test if this formatted name matches the "match" regex which in
+     * this case is "(Admin\.(.*))".  Meaning anything that begins with "Admin".
+     *
+     * It does, so then we move on to the replacement regex:
+     * "s/Admin\.(.*)/$1/".
+     *
+     * The "Admin\.(.*)" is the from pattern.
+     * The "$1" is the to pattern.
+     *
+     * This rule effectively removes the "Admin." at the beginning of the
+     * formatted name:
+     * "site-userfooexample.com" is the result.
+     */
+
+    ss::sstring result;
+
+    if (_is_default) {
+        if (_default_realm == params[0]) {
+            result = ss::sstring(params[1]);
+        }
+    } else if (
+      params.size() - 1 == static_cast<size_t>(_number_of_components)) {
```

Comment:
there probably needs to be some checking some where that params conforms to correct input. some branches it needs to be len()>=2 some needs to be non-empty, etc...